saleor/order/views.py

```diff
@@ -82,52 +87,34 @@ def payment(request, token):
 
 @check_order_status
 def start_payment(request, order, variant):
-    waiting_payments = order.payments.filter(
-        status=PaymentStatus.WAITING).exists()
-    if waiting_payments:
-        return redirect('order:payment', token=order.token)
-    billing = order.billing_address
-    total = order.total
     defaults = {
-        'total': total.gross.amount,
-        'tax': total.tax.amount,
-        'currency': total.currency,
-        'delivery': order.shipping_price.net.amount,
-        'billing_first_name': billing.first_name,
-        'billing_last_name': billing.last_name,
-        'billing_address_1': billing.street_address_1,
-        'billing_address_2': billing.street_address_2,
-        'billing_city': billing.city,
-        'billing_postcode': billing.postal_code,
-        'billing_country_code': billing.country.code,
-        'billing_email': order.user_email,
-        'description': pgettext_lazy(
-            'Payment description', 'Order %(order_number)s') % {
-                'order_number': order},
-        'billing_country_area': billing.country_area,
-        'customer_ip_address': get_client_ip(request)}
-    variant_choices = settings.CHECKOUT_PAYMENT_CHOICES
-    if variant not in [code for code, dummy_name in variant_choices]:
+        'customer_ip_address': get_client_ip(request),
+        **get_billing_data(order)}
+    if variant not in settings.CHECKOUT_PAYMENT_CHOICES:
         raise Http404('%r is not a valid payment variant' % (variant,))
     with transaction.atomic():
-        payment, dummy_created = Payment.objects.get_or_create(
-            variant=variant, status=PaymentStatus.WAITING, order=order,
-            defaults=defaults)
-        try:
-            form = payment.get_form(data=request.POST or None)
-        except RedirectNeeded as redirect_to:
-            return redirect(str(redirect_to))
-        except Exception:
-            logger.exception('Error communicating with the payment gateway')
-            msg = pgettext_lazy(
-                'Payment gateway error',
-                'Oops, it looks like we were unable to contact the selected '
-                'payment service')
-            messages.error(request, msg)
-            payment.change_status(PaymentStatus.ERROR)
-            return redirect('order:payment', token=order.token)
+        payment, _ = Payment.objects.get_or_create(
+            variant=variant, is_active=True, order=order, defaults=defaults,
+            total=order.total.gross.amount, currency=order.total.gross.currency)
+        if (
+                order.is_fully_paid()
+                or payment.charge_status == ChargeStatus.FULLY_REFUNDED):
+            return redirect(order.get_absolute_url())
+        provider, provider_params = get_provider(payment.variant)
+        transaction_token = provider.get_transaction_token(**provider_params)
+        form = get_form_for_payment(payment)
+        form = form(data=request.POST or None, instance=payment)
```

Comment:
What do you think about passing the provider? Like that:
```suggestion
        form = form(data=request.POST or None, provider=provider, instance=payment)
```
So we could access data to construct the form (e.g.: the public key to put in the `django.forms.Form`'s `attrs` constructor parameter).

Reply:
Sure, if that would be helpful